packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -191,25 +205,59 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
   /// can override this method to customize appearance of text.
   TextSpan buildTextSpan({required BuildContext context, TextStyle? style , required bool withComposing}) {
     assert(!value.composing.isValid || !withComposing || value.isComposingRangeValid);
+
+    final Map<TextRange, InlineSpan> rangeSpanMapping= <TextRange, InlineSpan>{};
     // If the composing range is out of range for the current text, ignore it to
     // preserve the tree integrity, otherwise in release mode a RangeError will
     // be thrown and this EditableText will be built with a broken subtree.
-    if (!value.isComposingRangeValid || !withComposing) {
-      return TextSpan(style: style, text: text);
+    if (value.isComposingRangeValid && withComposing) {
+      // return TextSpan(style: style, text: text);
+      final TextStyle composingStyle = style!.merge(
+        const TextStyle(decoration: TextDecoration.underline),
+      );
+      rangeSpanMapping[value.composing] = TextSpan(
+        style: composingStyle,
+        text: value.composing.textInside(value.text),
+      );
     }
-    final TextStyle composingStyle = style!.merge(
-      const TextStyle(decoration: TextDecoration.underline),
-    );
-    return TextSpan(
+    for (final TextEditingInlineSpanReplacement replacement in textEditingInlineSpanReplacements) {
```

Comment:
while testing out the pr, i noticed the entire app would freeze and crash if deleting too fast. im now wondering if these fors could be the culprit, as 3 nested fors could be heavy. for solving this, i was thinking of using the setter for textEditingInlineSpanReplacements for compiling the rangeSpanMapping instead of building these when building the text spans

Reply:
This PR is now in a much more mature state. I'd be interested to see if you still see the freeze/crash. I am not seeing it on my test devices. The nested for loops are indeed inefficient for large amounts of text, but I would be very surprised to see it make a noticeable difference for the vast majority of use cases. We do similar/more work for the layout itself that this should be insignificant amount of work barring thousands of matched TextRanges.

Since the text can change at any point, the spans must be calculated on each text change event to ensure what we output at any given point is correct. Computing the replacements in the setter would not be able to capture all of the replacements as text is being typed.

Reply:
Also, keep in mind, debug mode performance is much worse than release mode performance and is not indicative of the actual performance of the app.

Reply:
Upon further testing, turns out the issue was on my app's code and not the pr itself. The pr seems to be working pretty good, gj!